lib/veri.ml

```diff
@@ -0,0 +1,237 @@
+      | Some data ->
+        self#update_event (create_reg_read var data) >>= fun () ->
+        SM.return r
+      | None -> SM.return r
+    else SM.return r
+
+  method! update var result : 'a u =
+    super#update var result >>= fun () -> 
+    if is_imm_var var then 
+      match imm_result (value result) with 
+      | Some data -> self#update_event (create_reg_write var data)
+      | None -> do_nothing ()
+    else do_nothing ()
+
+  method private eval_mem_event tag addr data : 'a e = 
+    match value addr, value data with
+    | Bil.Imm addr, Bil.Imm data ->
+      let ev = create_move_event tag addr data in
+      SM.return (Some ev)
+    | _ -> SM.return None
+
+  method! eval_event ev =
+    super#eval_event ev >>= fun () -> self#update_event ev
+      
+  method! eval_store: mem:exp -> addr:exp -> exp -> endian -> size -> 'a r =
+    fun ~mem ~addr data endian size ->
+      super#eval_store ~mem ~addr data endian size >>= fun r ->
+      self#eval_exp addr >>= fun addr ->
+      self#eval_exp data >>= fun data ->
+      self#eval_mem_event Event.memory_store addr data >>=
+      fun ev -> match ev with
+      | None -> SM.return r
+      | Some ev -> self#update_event ev >>= fun () -> SM.return r
+
+  method private resolve_addr : mem:exp -> addr:exp -> endian -> size -> 'a r = 
+    fun ~mem ~addr endian size -> 
+      self#eval_exp addr >>= fun addr_res -> 
+      match value addr_res with
+      | Bil.Bot | Bil.Mem _ -> super#eval_load ~mem ~addr endian size
+      | Bil.Imm addr' ->
+        SM.get () >>= fun ctxt ->
+        let same_addr mv = Move.cell mv = addr' in
+        match find_mem_load ctxt#real_events same_addr with
+        | None -> super#eval_load ~mem ~addr endian size 
+        | Some mv ->
+          let data = Bil.int (Move.data mv) in
+          super#eval_store ~mem ~addr data endian size >>= fun r ->
+          match value r with 
+          | Bil.Mem _ -> 
+            super#update mem_var r >>= fun () ->
+            super#eval_load ~mem ~addr endian size
+          | Bil.Imm _ | Bil.Bot -> return r
+
+  method! eval_load: mem:exp -> addr:exp -> endian -> size -> 'a r = 
+    fun ~mem ~addr endian size -> 
+      self#resolve_addr mem addr endian size >>= fun r ->
+      self#eval_exp addr >>= fun addr' ->
+      self#eval_mem_event Event.memory_load addr' r >>= fun ev ->
+      match ev with
+      | Some ev -> self#update_event ev >>= fun () -> SM.return r
+      | None -> SM.return r      
+
+  method private eval_chunk arch chunk = 
+    match Veri_helpers.insn_of_chunk arch chunk with
+    | Error er -> do_nothing () 
+    | Ok ((mem,insn) as insn') ->       
+      let name = Disasm_expert.Basic.Insn.name insn in
+      SM.update (fun c -> c#set_description name) >>= fun () ->
+      match Veri_helpers.lift_insn arch insn' with
+      | None -> SM.update (fun c -> c#update_report (`Unlif name))
+      | Some bil -> self#eval bil
+                      
+  method private step point = 
```

Comment:
the last two methods doesn't look nice for me. You're completely overriding the `eval_trace` method, instead of extending it. That rings, that you have some problem here.